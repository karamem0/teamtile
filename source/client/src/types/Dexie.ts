//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface ArrayEntity<T> {
  id: string,
  expired: number,
  values: T[]
}

export interface ValueEntity<T> {
  id: string,
  expired: number,
  value: T
}
