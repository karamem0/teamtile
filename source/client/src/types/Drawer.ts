//
// Copyright (c) 2021-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export type DrawerType =
  | 'channel'
  | 'member'
  | 'tag';
