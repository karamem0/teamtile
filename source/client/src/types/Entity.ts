//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export enum MembershipType {
  private = 'private',
  standard = 'standard'
}

export enum VisibilityType {
  private = 'private',
  public = 'public'
}

export enum AccordionType {
  all = 'all',
  archived = 'archived',
  pinned = 'pinned'
}

export interface Channel {
  id?: string,
  displayName?: string,
  membershipType?: MembershipType,
  webUrl?: string
}

export interface Drive {
  id?: string,
  webUrl?: string
}

export interface Group {
  id?: string,
  email?: string,
  sensitivityLabel?: string
}

export interface Icon {
  id?: string,
  data?: string
}

export interface Member {
  id?: string,
  displayName?: string,
  email?: string,
  icon?: string,
  userId?: string
}

export interface Tab {
  id?: string,
  appId?: string,
  displayName?: string,
  webUrl?: string
}

export interface Team {
  id?: string,
  archived?: boolean,
  description?: string,
  displayName?: string,
  icon?: string,
  internalId?: string,
  visibility?: VisibilityType,
  webUrl?: string
}

export interface TeamInfo {
  id?: string,
  displayName?: string,
  tenantId?: string
}
