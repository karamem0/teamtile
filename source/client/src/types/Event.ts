//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

export type Event = React.SyntheticEvent | React.SyntheticEvent<unknown> | Record<string, never>;

export type EventHandler<T = never> = (
  event?: Event,
  data?: T
) => void;
