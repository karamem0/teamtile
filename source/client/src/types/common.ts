//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface KeyValue<K, V> {
  key: K,
  value: V
}
