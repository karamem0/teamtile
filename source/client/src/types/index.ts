//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface Team {
  id?: string;
  internalId?: string;
  name?: string;
  description?: string;
  visibility?: string;
  url?: string;
  icon?: string;
}

export interface TeamChannels {
  count?: number;
  nextLink?: string;
  values?: TeamChannel[];
}

export interface TeamChannel {
  id?: string;
  name?: string;
  url?: string;
}

export interface TeamMembers {
  count?: number;
  nextLink?: string;
  values?: TeamMember[];
}

export interface TeamMember {
  id?: string;
  name?: string;
  email?: string;
}

export interface TeamDrive {
  id?: string;
  url?: string;
}
