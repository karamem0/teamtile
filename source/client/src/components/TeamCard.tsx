import React from 'react';
import * as microsoftTeams from '@microsoft/teams-js';
import {
  Button,
  Card,
  Text
} from '@fluentui/react-northstar';
import { LockIcon, GlobeIcon } from '@fluentui/react-icons';
import { Team } from '../types';
import AppContext from '../contexts/AppContext';
import useTeam from '../hooks/useTeam';
import TeamIcon from './TeamIcon';
import TeamMenu from './TeamMenu';

interface TeamCardProps {
  team?: Team;
}

const TeamCard: React.FC<TeamCardProps> = (props: TeamCardProps) => {

  const [ token, , , setError ] = React.useContext(AppContext);
  const [
    team,
    channels,
    members,
    drive,
    error
  ] = useTeam({ token, team: props.team });

  React.useEffect(() => {
    if (!setError) {
      return;
    }
    setError(error);
  }, [ setError, error ]);

  return (
    <React.Fragment>
      {
        team &&
          <Card
            className="card"
            fluid>
            <div className="card-column">
              <div className="card-column-item">
                <TeamIcon
                  icon={team.icon}
                  name={team.name} />
              </div>
              <div className="card-column-item">
                <div className="card-row">
                  <Button
                    className="card-name"
                    content={team.name}
                    text
                    onClick={() => {
                      if (!team.url) {
                        return;
                      }
                      microsoftTeams.executeDeepLink(team.url);
                    }} />
                  <Text
                    className="card-description"
                    content={team.description}
                    size="small"
                    truncated />
                  <TeamMenu
                    channels={channels}
                    drive={drive}
                    members={members}
                    team={team} />
                </div>
              </div>
              <div className="card-column-item">
                <Text color="brand">
                  {
                    team.visibility === 'public' &&
                      <GlobeIcon className="card-menu-icon" />
                  }
                  {
                    team.visibility === 'private' &&
                      <LockIcon className="card-menu-icon" />
                  }
                </Text>
              </div>
            </div>
          </Card>
      }
    </React.Fragment>
  );

};

export default TeamCard;
