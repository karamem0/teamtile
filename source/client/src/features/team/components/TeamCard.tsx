//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { app } from '@microsoft/teams-js';

import { Item } from '../../../types/Store';

import Presenter from './TeamCard.presenter';

export interface TeamCardProps {
  item: Item
}

function TeamCard(props: TeamCardProps) {

  const { item } = props;

  const handleClick = React.useCallback(() => {
    if (!item.value.webUrl) {
      return;
    }
    app.openLink(item.value.webUrl);
  }, [ item ]);

  return (
    <Presenter
      item={item}
      onClick={handleClick} />
  );

}

export default TeamCard;
