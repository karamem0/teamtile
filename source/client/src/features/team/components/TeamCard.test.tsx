//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { render, screen } from '@testing-library/react';

import { VisibilityType } from '../../../types/Entity';

import TeamCard from './TeamCard.presenter';

jest.mock('./CalendarMenuItem', () =>
  function CalendarMenuItem({ children }: React.PropsWithChildren<unknown>) {
    return (
      <div data-testid="CalendarMenuItem">
        {children}
      </div>
    );
  });

jest.mock('./ChannelMenuItem', () =>
  function ChannelMenuItem({ children }: React.PropsWithChildren<unknown>) {
    return (
      <div data-testid="ChannelMenuItem">
        {children}
      </div>
    );
  });

jest.mock('./DriveMenuItem', () =>
  function DriveMenuItem({ children }: React.PropsWithChildren<unknown>) {
    return (
      <div data-testid="DriveMenuItem">
        {children}
      </div>
    );
  });

jest.mock('./MemberMenuItem', () =>
  function MemberMenuItem({ children }: React.PropsWithChildren<unknown>) {
    return (
      <div data-testid="MemberMenuItem">
        {children}
      </div>
    );
  });

test('create shapshot of when item loading is true', async () => {
  const params = {
    item: {
      id: '02bd9fd6-8f93-4758-87c3-1fb73740a315',
      value: {
        id: '02bd9fd6-8f93-4758-87c3-1fb73740a315',
        displayName: 'HR Taskforce',
        description: 'Welcome to the HR Taskforce team.',
        internalId: '19:09fc54a3141a45d0bc769cf506d2e079@thread.skype',
        sensitivityLabel: 'Restricted',
        visibility: 'private' as VisibilityType,
        webUrl: 'https://teams.microsoft.com/l/team/19:09fc54a3141a45d0bc769cf506d2e079%40thread.skype/conversations?groupId=02bd9fd6-8f93-4758-87c3-1fb73740a315&tenantId=dcd219dd-bc68-4b9b-bf0b-4a33a796be35'
      },
      loading: false,
      visible: true
    }
  };
  render(<TeamCard {...params} />);
  expect((await screen.findAllByText(/^.*$/))[0]).toMatchSnapshot();
});

test('create shapshot of when item loading is false', async () => {
  const params = {
    item: {
      id: '02bd9fd6-8f93-4758-87c3-1fb73740a315',
      value: {
        id: '02bd9fd6-8f93-4758-87c3-1fb73740a315',
        displayName: 'HR Taskforce',
        description: 'Welcome to the HR Taskforce team.',
        internalId: '19:09fc54a3141a45d0bc769cf506d2e079@thread.skype',
        sensitivityLabel: 'Restricted',
        visibility: 'private' as VisibilityType,
        webUrl: 'https://teams.microsoft.com/l/team/19:09fc54a3141a45d0bc769cf506d2e079%40thread.skype/conversations?groupId=02bd9fd6-8f93-4758-87c3-1fb73740a315&tenantId=dcd219dd-bc68-4b9b-bf0b-4a33a796be35'
      },
      loading: true,
      visible: true
    }
  };
  render(<TeamCard {...params} />);
  expect((await screen.findAllByText(/^.*$/))[0]).toMatchSnapshot();
});
