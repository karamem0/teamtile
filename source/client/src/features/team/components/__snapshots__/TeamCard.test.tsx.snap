// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create shapshot of when item loading is false 1`] = `
<DocumentFragment>
  .emotion-0 {
  min-height: 100vh;
  line-height: 1.25em;
  background-color: #f5f5f5;
}

.emotion-1 {
  height: 6rem;
}

.emotion-2 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  grid-gap: 0.5rem;
}

.emotion-3 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto;
  grid-gap: 0.25rem;
}

<div
    class="fui-FluentProvider fui-FluentProviderr1 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <div
        class="fui-Card emotion-1 ___139ns38_k98ud00 f1a3p1vp flpxrij fi54ll6 fs0zar4 f22iagw f10pi13n f1ewtqcl f19n0e5 f1mdlcz9 frwkxtg f1n6gb5g f15yvnhg fo72kxq f13zj6fq f1nwj1ja f8rth92 flthirb f1lh990p ftkbnf5 f6czdpx f13hvwk3 f1i1u9k0 f1en4csx fw3u9gn fnk4z59 f1vx9l62 f14k419y f1fgo9fz fvqmfsm f3am6yf f1r5wgso frsmuga fuldkky fxugw4r f1whvlc6 f16gxe2i fpgykix f1osi826 fzybk4o ftqa4ok f2hkw1w f8hki3x f1d2448m f1bjia2o ffh67wi f226i61 f13kzufm flujwa2 fsx75g8 f15bsgw9 f14e48fq f18yb2kv fd6o370 fpqizxz fnd8nzh f15fr7a0 fwsq40z f34ld9f fy0y4wt"
        role="listitem"
      >
        <div
          aria-busy="true"
          aria-label="Loading Content"
          class="fui-Skeleton"
          role="progressbar"
        >
          <div
            class="emotion-2"
          >
            <div
              class="fui-SkeletonItem ___18ygo6q_qogacp0 f10pi13n f1a3p1vp f1qx3921 fj9j8l8 f1b6djjb f1dsdmen f1cpbl36 f1iuewzk f452v7t f4akx1t f18p5put fnm2mpv f1gjxg63 f5q2cvs f124akge ff2sm71"
            />
            <div
              class="emotion-3"
            >
              <div
                class="fui-SkeletonItem ___h30iyh0_yafwio0 f10pi13n f1a3p1vp f1qx3921 fj9j8l8 f1b6djjb f1dsdmen f1cpbl36 f1iuewzk f452v7t f4akx1t f18p5put fnm2mpv f1gjxg63 fly5x3f ff3glw6 frvgh55"
              />
              <div
                class="fui-SkeletonItem ___oybdil0_1t3zkkb f10pi13n f1a3p1vp f1qx3921 fj9j8l8 f1b6djjb f1dsdmen f1cpbl36 f1iuewzk f452v7t f4akx1t f18p5put fnm2mpv f1gjxg63 fly5x3f ff3glw6 fjamq6b"
              />
              <div
                class="fui-SkeletonItem ___oybdil0_1t3zkkb f10pi13n f1a3p1vp f1qx3921 fj9j8l8 f1b6djjb f1dsdmen f1cpbl36 f1iuewzk f452v7t f4akx1t f18p5put fnm2mpv f1gjxg63 fly5x3f ff3glw6 fjamq6b"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`create shapshot of when item loading is true 1`] = `
<DocumentFragment>
  .emotion-0 {
  min-height: 100vh;
  line-height: 1.25em;
  background-color: #f5f5f5;
}

.emotion-1 {
  height: 6rem;
}

.emotion-2 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr;
  grid-gap: 0.5rem;
}

.emotion-3 {
  display: grid;
  grid-gap: 0.25rem;
}

.emotion-4 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: 1fr auto auto;
  grid-gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-6 {
  cursor: pointer;
}

.emotion-7 {
  font-weight: 600;
  color: #4c6978;
}

.emotion-9 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto auto auto auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -ms-flex-pack: left;
  -webkit-justify-content: left;
  justify-content: left;
}

<div
    class="fui-FluentProvider fui-FluentProviderr0 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <div
        class="fui-Card emotion-1 ___139ns38_k98ud00 f1a3p1vp flpxrij fi54ll6 fs0zar4 f22iagw f10pi13n f1ewtqcl f19n0e5 f1mdlcz9 frwkxtg f1n6gb5g f15yvnhg fo72kxq f13zj6fq f1nwj1ja f8rth92 flthirb f1lh990p ftkbnf5 f6czdpx f13hvwk3 f1i1u9k0 f1en4csx fw3u9gn fnk4z59 f1vx9l62 f14k419y f1fgo9fz fvqmfsm f3am6yf f1r5wgso frsmuga fuldkky fxugw4r f1whvlc6 f16gxe2i fpgykix f1osi826 fzybk4o ftqa4ok f2hkw1w f8hki3x f1d2448m f1bjia2o ffh67wi f226i61 f13kzufm flujwa2 fsx75g8 f15bsgw9 f14e48fq f18yb2kv fd6o370 fpqizxz fnd8nzh f15fr7a0 fwsq40z f34ld9f fy0y4wt"
        role="listitem"
      >
        <div
          class="emotion-2"
        >
          <div
            data-testid="AvatarIcon"
          />
          <div
            class="emotion-3"
          >
            <div
              class="emotion-4"
            >
              <div
                class="emotion-5"
              >
                <span
                  class="fui-Text emotion-6 ___c56hd20_1xy0w11 fk6fouc fkhj508 f1i3iumi figsok6 fpgzoln f1w7gpdv f6juhto f1gl81tg f2jf649"
                  role="button"
                >
                  <span
                    class="fui-Text emotion-7 ___c56hd20_1xy0w11 fk6fouc fkhj508 f1i3iumi figsok6 fpgzoln f1w7gpdv f6juhto f1gl81tg f2jf649"
                  >
                    HR Taskforce
                  </span>
                </span>
              </div>
              <div
                data-testid="SensitivityLabel"
              />
              <div
                data-testid="VisibilityIcon"
              />
            </div>
            <div
              class="emotion-5"
            >
              Welcome to the HR Taskforce team.
            </div>
            <div
              class="emotion-9"
            >
              <div
                data-testid="ChannelMenuItem"
              />
              <div
                data-testid="MemberMenuItem"
              />
              <div
                data-testid="DriveMenuItem"
              />
              <div
                data-testid="CalendarMenuItem"
              />
              <div
                data-testid="PinMenuItem"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
